File: src/test/java/edu/wpi/teamname/views/HomeTest.java
package edu.wpi.teamname.views;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;
import static org.testfx.api.FxAssert.verifyThat;

import edu.wpi.teamname.services.ServiceTwo;
import edu.wpi.teamname.services.database.DatabaseService;
import edu.wpi.teamname.state.HomeState;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import javafx.fxml.FXMLLoader;
import javafx.scene.Node;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.stage.Stage;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Spy;
import org.mockito.junit.jupiter.MockitoExtension;
import org.testfx.api.FxRobot;
import org.testfx.framework.junit5.ApplicationExtension;
import org.testfx.framework.junit5.Start;

@ExtendWith({MockitoExtension.class, ApplicationExtension.class})
public class HomeTest extends FxRobot {

  // Mock dependencies
  @Mock ServiceTwo two;
  @Mock DatabaseService dbService;
  @Mock FXMLLoader mockloader;
  // Don't mock since we require the state to change
  @Spy HomeState state;

  @InjectMocks HomeController controller;

  private final FXMLLoader loader = new FXMLLoader();
  private Parent sceneRoot;

  @Start
  private void start(Stage stage) throws IOException {
    loader.setControllerFactory((i) -> controller);
    sceneRoot = loader.load(getClass().getResourceAsStream("HomeView.fxml"));
    var primaryScene = new Scene(sceneRoot);
    // We are forced to inject this object afterward, since it must be created on the JavaFX thread
    controller.setAppPrimaryScene(primaryScene);

    stage.setScene(primaryScene);
    stage.setAlwaysOnTop(true);
    stage.show();
  }

  @BeforeEach
  public void init() throws IOException {
    when(mockloader.load(any(InputStream.class))).thenReturn(sceneRoot);
    when(dbService.getEmployeeName()).thenReturn("Wilson Wong");
  }

  @Test
  public void testSingleClick() {
    when(two.getResults()).thenCallRealMethod();
    clickOn("Click Me");
    verifyThat("#text", Node::isVisible);
  }

  @Test
  public void testDoubleClick() {
    when(two.getResults()).thenCallRealMethod();
    clickOn("Click Me");
    clickOn("Click Me");
    verifyThat("#text", (n) -> !n.isVisible());
  }

  @Test
  public void testGetNodes() {
    when(two.getResults())
        .thenReturn(
            new ArrayList<>() {
              {
                add("Expected");
              }
            });
    clickOn("Click Me");
    verifyThat("Expected", Node::isVisible);
  }
}

File: src/test/kotlin/edu/wpi/teamname/views/HomeTest.kt
package edu.wpi.teamname.views

import edu.wpi.teamname.services.ServiceTwo
import edu.wpi.teamname.services.database.DatabaseService
import edu.wpi.teamname.state.HomeState
import javafx.fxml.FXMLLoader
import javafx.scene.Node
import javafx.scene.Scene
import javafx.scene.layout.AnchorPane
import javafx.stage.Stage
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.kotlin.any
import org.mockito.kotlin.given
import org.mockito.kotlin.mock
import org.testfx.api.FxAssert
import org.testfx.api.FxRobot
import org.testfx.framework.junit5.ApplicationExtension
import org.testfx.framework.junit5.Start

@ExtendWith(ApplicationExtension::class)
class HomeTest : FxRobot() {
    // Mock dependencies
    private val two: ServiceTwo = mock {
        given(it.results).willReturn(listOf("Results 1", "Results 2"))
    }

    private var dbService: DatabaseService = mock {
        given(it.employeeName).willReturn("Wilson Wong")
    }

    private val mockLoader: FXMLLoader = mock()

    private val state = HomeState()

    private val controller: HomeController = HomeController(dbService, two, mockLoader, state)
    private val loader: FXMLLoader = FXMLLoader()

    @Start
    private fun start(stage: Stage) {
        loader.setControllerFactory { controller }
        val sceneRoot: AnchorPane = loader.load(javaClass.getResourceAsStream("HomeView.fxml"))
        given(mockLoader.load<AnchorPane>(any())).willReturn(sceneRoot)
        val primaryScene = Scene(sceneRoot)
        stage.scene = primaryScene
        stage.isAlwaysOnTop = true
        stage.show()
    }

    @Test
    fun testSingleClick() {
        clickOn("Click Me")
        FxAssert.verifyThat("#text") { obj: Node -> obj.isVisible }
    }

    @Test
    fun testDoubleClick() {
        clickOn("Click Me")
        clickOn("Click Me")
        FxAssert.verifyThat("#text") { n: Node -> !n.isVisible }
    }

    @Test
    fun testGetNodes() {
        given(two.results).willReturn(listOf("Expected", "Not expected"))
        clickOn("Click Me")
        FxAssert.verifyThat("Wilson Wong, Expected") { obj: Node -> obj.isVisible }
    }
}
